# Data
## Introduction
In empirical research for quantitative policy analysis, data takes a central
stage. Becoming conversant in the proper treatment of data is thus a vital skill
to a public policy researcher. This chapter begins with the distinction between
data, information, knowledge, and wisdom.  Data comes in
a rich variety of forms, and we provide an overview of the main forms of
data sets used in empirical research. We then move on to cover the different
measurement scales that are encountered in quantitative analysis. With these conceptual matters under our
belts, we then consider various mathematical operations and transformations that
are frequently applied in public policy analysis.
The chapter then covers the reading and processing of external data sets.

## Data, Information, Knowledge, Wisdom

While data will be central stage in this course, it is important to view data
from a larger perspective. In the so-called DIKW-Pyramid, data sits at the
bottom of a hierarchy often encountered in information science {cite}`Wallace_2007`. 
From this perspective, data is viewed as discrete or objective facts that
remain unorganized and lacking context or information.

<a title="Longlivetheux / CC BY-SA (https://creativecommons.org/licenses/by-sa/4.0)" href="https://commons.wikimedia.org/wiki/File:DIKW_Pyramid.svg"><img width="256" alt="DIKW Pyramid" src="https://upload.wikimedia.org/wikipedia/commons/thumb/0/06/DIKW_Pyramid.svg/256px-DIKW_Pyramid.svg.png"></a>

At the next step of the hierarchy, when data is imbued with meaning, we arrive
at *information*. We might think of information as describing *what* is in
the data. Further on up the pyramid is knowledge, representing the subjective
perception of the world as seen through the lower levels of the pyramid.
We might think of knowledge as our individual understanding of how things
relate to one another.
Finally, we arrive at wisdom, which is sometimes described as "knowing the
right things to do". From the perspective of this class, policy would live at
this final level.


## Data Sets 

A **data set** is a collection of **observations** recorded for individual **units** on a set of **variables**.
Variables are sometimes referred to as **attributes** or **features** (in machine
learning parlance).


To familiarize ourselves with data sets, we will use an
empirical example from COVID case and death counts, collected by the
New York Times and made available on their [COVID githhub
repository](https://github.com/nytimes/covid-19-data ). This file was
downloaded on 2020-08-02 and saved into the book repository.


### CSV Format

The file is a *csv* or comma separated file. To see what this file format entails, we can
open a terminal in JupyterLab and navigate to the directory `notebooks/data/covid`
using the `cd` command. Following this, the Unix command `head` can be called
to print out the first 5 lines of the file:


```{figure} ../figs/data/covid_head_shell.png
---
name: f:covid_head_shell
---
Inspecting a csv file in the shell
```

```{note} 
AUTHOR: Change the example figure here to have the directory structure setup on the
jupyterhub instance on a student account.
```
This kind of introspection at the shell can be very useful in trying to
understand the layout and contents of a file. The results of the `head` command
reveal that the first line of the file is a *header* where a series of strings,
or variable names, are separated by a comma `,`. The comma is referred to as a
*delimiter*.

The subsequent lines in the file contain the first four records in the data
set. Again the values for the variables in a given row are separated by the
same delimiter used in the header row.

We can learn more about our file at the shell. We might want to know how many
lines are in the file. The `wc` (wc is short for 'word count') command comes to our aide here:

```
☁  covid [master] ⚡ wc -l covid_states_20200802.csv
8430 covid_states_20200802.csv
☁  covid [master] ⚡ 
```

The 8430 lines in the file are composed of the 1 header line and 8429 record lines.

### Why not Excel?

You may have encountered the csv format if you have any experience working with
spreadsheets, such as Microsoft Excel. It is important to note, however, that
spreadsheets are the effect and csv are the cause here. Indeed, the csv format
predates the personal computer (and spreadsheets) and was developed to support
the exchange of tabular data between mainframe database programs.[^csv]

[^csv]: For the history of comma separated value formats see [Comma-separated values](https://en.wikipedia.org/wiki/Comma-separated_values).

If you have used a spreadsheets program, you may be wondering why we are not using one to
read this file, since they were designed for such data. There are a
number of reasons for not using spreadsheets in quantitative analysis. The
first has to do with reproducible science, which we encountered in the introduction
to this book. File formats for spreadsheets are often proprietary which requires
specific software for reading the binary format. These formats are ill-suited
for use with code repository systems (GitHub, Bitbucket, GitLab) that are the
foundation of modern reproducible science. Moreover, those proprietary formats are
subject to change overtime and this can further break reproducibility.


The second set of reasons for using computational scripting rather than Excel
for quantitative analysis in public policy is that the scripting approach is
much more flexible, scalable, and transparent. By flexible, we mean that
scripts can accomplish, in a few lines, what would take complex macros inside a
spreadsheet. Scalable means that computational scripting fits a wide range of
data sets, from small (i.e., they can fit inside the memory of our laptops) to
massive data sets that can only be analyzed on clusters and high-performance
infrastructures. Spreadsheets are not used in the latter cases, while scripts
are the *lingua franca* of the big data world. Finally, the open nature of the
scripts makes the scientific and policy analysis workflows transparent. There
is no black box here. In contrast, when using a complex spreadsheet, formulae
are often buried inside of cells, or even on different sheets/pages, and this
can obscure the understanding of the analysis.
<!--- The concept of the 'black box' and the problems from a public policy perspective were a big part of my learning from a lesson about the EPA in PBPl 200, which was taught concurrently with 210, so this is good to bring up/emphasize> <--->
Finally, spreadsheets have been behind some of the most prominent cases of
errors in data analysis in recent history. These errors have had massive
implications for public policy. The poster child is the so-called 'Excel
Depression' where a mistake in a spreadsheet was found to undermine the
quantitative support for fiscal austerity programs that were imposed across the
world during the Great Recession {cite}`Herndon_2013`. This is not to say that
mistakes cannot be made in working with computational scripting. They can, and
they are. However, the flexible nature of the scripting workflows and the open
file formats that they employ make the detection, and correction, of such
errors much more likely than when working in spreadsheets.


### CSV with JupyterLab

Moreover, because JupyterLab leverages Python and Unix under the hood, its
capabilities vastly outstrip what one can do with a data set using a
spreadsheet. We've already seen we can introspect the file using a terminal in
JupyerLab, but we can also open the file directly: 

```{figure} ../figs/data/covid_csv.png
---
name: f:data/covid
---
State COVID Data from the New York Times
```
The records are daily, since the beginning of reported cases in the US, and
continue up until the most recent date (August 2). Looking at the first few
records shown in the figure, we see that not all states are included initially.
However, if we jump to the bottom of the file we see:

```{figure} ../figs/data/covid_tail.png
---
name: f:data/covid_tail
---
State COVID Data from the New York Times (Bottom of file)
```
so by August 2, all states have been included in the data set.


## Pandas
For processing this data set we will rely on 
the Python package [pandas](http://pandas.org) for reading and manipulating
this data set.

In [5]:
import pandas

The `import pandas` expression brings the pandas package into the *namespace* of
the interpreter. This gives us access to the functionality of the package. 

### Reading files with Pandas

Once we have pandas in memory we can read the COVID data file:

In [6]:
df = pandas.read_csv('data/covid/covid_states_20200802.csv')

`read_csv` is a function in the pandas package that reads csv
files. As an argument to function we pass in the *file path*. This is a string
that gives pandas information about where the file is located. This is a
relative path, meaning in the current directory (where the notebook is being
run) there is a directory called `data`, inside `data` there is a second directory called `covid` and that directory contains the file
`covid_states_20200802.csv`.

### DataFrame

The result of the function is assigned into a variable called `df`.
We can examine what data type this variable is:

In [7]:
type(df)

pandas.core.frame.DataFrame

Pandas DataFrames are very powerful for data manipulation. We will be
illustrating some of this functionality in order to explore this data set.

We can examine the top of the file:

In [8]:
df.head()

date       state  fips  cases  deaths
0  2020-01-21  Washington    53      1       0
1  2020-01-22  Washington    53      1       0
2  2020-01-23  Washington    53      1       0
3  2020-01-24    Illinois    17      1       0
4  2020-01-24  Washington    53      1       0

````{margin}
```{note}
If we wanted to see more rows of data, we can simply pass an argument inside the parenthesis (). For example, if we wanted to see the first ten rows of data for our COVID DataFrame, we would add a 10 inside  **df.head()**: `df.head(10)`. You could repeat this with tail() as well. 

```
````

Alternatively, we can also see the end of the file:

In [9]:
df.tail()

date          state  fips  cases  deaths
8424  2020-08-02       Virginia    51  91782    2218
8425  2020-08-02     Washington    53  60161    1680
8426  2020-08-02  West Virginia    54   6854     117
8427  2020-08-02      Wisconsin    55  58990     956
8428  2020-08-02        Wyoming    56   2808      26

The size of the DataFrame can be determined from its `shape` attribute:

In [10]:
df.shape

(8429, 5)

The first number is the number of rows, while the second reports the number
of columns in the DataFrame. For now we can think of the rows as records, and
the columns as attributes or variables that we want to analyze from a
statistical point of view. Thus we see there are 8429 records with measurements on  5 variables.

As we will see in subsequent chapters, the term *variable* can take on multiple meanings depending on the context. Here, where we are talking about the columns of a data set, variable is used to denote an attribute, or field, of the data set. Later, we will repurpose the phrase in the context of a *random variable* and focus on the statistical distribution of values in that column.

The idea of distinguishing between a variable in a data set as a way to organize observations on some attribute, versus a statistical model that summarizes the probabilistic distribution of the values for that variable is an important distinction.




### Series

Each column in the DataFrame contain a pandas **Series**.
A pandas series is a one-dimensional array that is capable of holding data of
any type (float, integer, string, python objects).  Let’s examine one of the
series in our COVID DataFrame:

In [11]:
df.deaths

0          0
1          0
2          0
3          0
4          0
        ... 
8424    2218
8425    1680
8426     117
8427     956
8428      26
Name: deaths, Length: 8429, dtype: int64

We see two columns here, the first is the *index* which has a unique value for
each record in the series (and DataFrame). We will return to the index below.
The second column holds the values for this particular series.

After all the values, comes `Name: deaths, Length: 8429, dtype: int64`, these
are the name, number of values, and the data type of the series.

The data type of a series determines how the series is being encoded and what
mathematical operations are appropriate for the series. More generally
speaking, the data type is related to the concept of the *measurement scale*
for the series.


### Simple Access to DataFrame and Series
DataFrame and Series support very flexible accessing method as we will see in the following subsection.

#### Accessing an entry in Series
The methods we learned for accessing elements in a `list` or `str` can also be
applied in accessing Series.

For example, if we want to access the third entry in a given Series `s`, we
could write `s[2]`. We could also access an entry in a Series by using the index
associated with this entry. For example, if the third entry in the Series `s`
is indexed by *"index_3"*, we could access this entry by using `s['index_3']`
as well. Here we can think about `index_3` as the _index label_.

There is one interesting thing to notice. A situation where we might run into
trouble is that when the Series is indexed by integer. In this situation, does
`s[0]` refer to the first entry in `s`,  or to the entry with index *0* in the
series? The answer is it returns the entry with index *0*. If the parameter we
pass matches some index, the interpreter always gives a prioirty to index
accessing. However, in real cases, most indexes are not integers and even if
they are, they appear the same as the sequence in which they appear, just as
you see in the Cov-19 dataset.(Usually if the row index is not specified, the
default index becomes the sequence of each record in the dataset, starting from
0)


#### Accessing columns in DataFrame

If we want to access a column in a given DataFrame, say if we want to access
the column `cases` from the dataframe `df` that we initialized above, we could
do either `df.cases` or `df['cases']`, both of which return a Series object
that reveals the information in the `cases` column. The index for this Series
object is exactly the same as the row index in the original DataFrame.

It appears to be slightly different if we want to access multiple columns from
the DataFrame at one time. Supppose we want to access both the `cases` column
and the `deaths` column, we could do the following:

In [12]:
df[['deaths','cases']]

deaths  cases
0          0      1
1          0      1
2          0      1
3          0      1
4          0      1
...      ...    ...
8424    2218  91782
8425    1680  60161
8426     117   6854
8427     956  58990
8428      26   2808

[8429 rows x 2 columns]

We could pass a list that contains the names of the columns that we want to
access. The difference between accessing one column and accessing multiple
columns is that, the parameter that we pass when accessing one column is a
string, indicating the name of the column that we would like to access, and the
parameter that we pass when accessing multiple columns is a list of strings,
including the names of all the columns that we want to access.

Another difference is that, accessing one column returns a Series object and
accessing multiple columns returns a DataFrame object. This is straightforward
as Series is one-dimensional and DataFrame is two-dimensional. Accessing one
column results in a one-dimensional data structure and accessing multiple
columns results in a two-dimensional data structure.

One more thing to note here is, when accessing multiple columns, the order of
the columns names in the list match the order of the columns in the returned
DataFrame.

#### Accessing rows in DataFrame
Accessing rows in a DataFrame can also use `loc` and `iloc`. We use `iloc` when
the row is recognized by sequence and use `loc` when the row is recognized by
row index. For example, in we want to access the first row, we could do

In [13]:
df.iloc[[0]]

date       state  fips  cases  deaths
0  2020-01-21  Washington    53      1       0

Since the index of the first row is also 0, which means we could also use

In [14]:
df.loc[[0]]

date       state  fips  cases  deaths
0  2020-01-21  Washington    53      1       0

If we want to access multiple rows, say, the second, the third , and the fifth rows, we could do

In [15]:
df.iloc[[1,2,4]]

date       state  fips  cases  deaths
1  2020-01-22  Washington    53      1       0
2  2020-01-23  Washington    53      1       0
4  2020-01-24  Washington    53      1       0

Note that if we want to access multiple rows that are continuous on the dataset, we could use `:`. For example, if we want to aceess from the first to the fifth row, we could do

In [16]:
df.iloc[0:4]

date       state  fips  cases  deaths
0  2020-01-21  Washington    53      1       0
1  2020-01-22  Washington    53      1       0
2  2020-01-23  Washington    53      1       0
3  2020-01-24    Illinois    17      1       0

#### More general usage of iloc and loc
In general, `iloc` and `loc` could be used to access any portion of the dataset
in a very flexible way. `loc` is the abbreviation for location, which takes the
name of the row and column as the parameter, `iloc` is the abbreviaion for
index location, which takes the sequence of rows and columns as the parameter.
It make help to think about `loc` as refereing to _label_ positions in the index and `iloc` as
referring to integer positions in the index.

The `iloc` function could be interpreted as `iloc[row name(s) , column
name(s)]`, which means there are two parameters, the first is row names and the
second is column names. If there is only one parameter, then it is taken as row
names in default and retrieve all the columns. In the above example where we
use `iloc` and `loc` to access rows, only one paramter is passed. Consequently,
the function takes the parameter as the row information.

Similarly, the `loc` could be interpreted as `loc[row sequence(s) , column sequence(s)]`.

Both the `iloc` and the `loc` could be used to retrieve any entry, rows,
columns, slicing the dataset and etc.

If only one row and one column are specified in `iloc` or `loc`, then the
result is just one entry. For example, if we want to access the entry in the
second row and the third column, we could do

In [17]:
df.iloc[1,2]

53

Since the row index and the column index for the entry in the seoncd row and
the thrid column are 1 and 'fips', respectively. We could also access the same
data by using

In [18]:
df.loc[1,'fips']

53

If we want to access from the second rows to the fourth rows with columns `deaths` and `fips`, we could do

In [19]:
df.loc[1:4,['deaths' , 'fips']]

deaths  fips
1       0    53
2       0    53
3       0    17
4       0    53

 


## Measurement Scales

Conventionally, quantitative data is organized along four categories of a
measurement scale. One can think that the amount of information increases as we
move up the measurmenet scale.

### Nominal

An attribute that is encoded using tags or labels is measured on a nominal
scale. Although integers can serve as the labels, the integers have no
numerical value other than denoting membership in a different group.

We have two nominal series in our DataFrame: `state` and `fips`. The former
takes on string values:

In [20]:
df.state

0          Washington
1          Washington
2          Washington
3            Illinois
4          Washington
            ...      
8424         Virginia
8425       Washington
8426    West Virginia
8427        Wisconsin
8428          Wyoming
Name: state, Length: 8429, dtype: object

point out what object means here  # Need to finish this sentence


While the latter are integers. `fips` are a categorizing system developed by the U.S. Census Bureau which are simply numerical values representing a geographical area. A full `fips` code has values coresponding to both state and county. Beyond acting as a label, they have no value.
> The fips code for California is 06.<br>
> The fips code for Riverside County is 065.<br>
> The full fips code therefore is 06065.<br>

In [21]:
df.fips

0       53
1       53
2       53
3       17
4       53
        ..
8424    51
8425    53
8426    54
8427    55
8428    56
Name: fips, Length: 8429, dtype: int64

Some datasets only include fips which are relevent for their geographical scale. In this case, only the state level indicator is included

Permissible mathematical operations for a nominal variable are limited to *mode*, *frequency*, and *proportions*.

For example, with pandas, if we wanted to get the frequency with which each
unique `state` value appears in the data we use:

In [22]:
df[['state', 'date']].groupby(by='state').count()

date
state                         
Alabama                    143
Alaska                     144
Arizona                    190
Arkansas                   145
California                 191
Colorado                   151
Connecticut                148
Delaware                   145
District of Columbia       149
Florida                    155
Georgia                    154
Guam                       141
Hawaii                     150
Idaho                      143
Illinois                   192
Indiana                    150
Iowa                       148
Kansas                     149
Kentucky                   150
Louisiana                  147
Maine                      144
Maryland                   151
Massachusetts              184
Michigan                   146
Minnesota                  150
Mississippi                145
Missouri                   149
Montana                    143
Nebraska                   168
Nevada                     151
New Hampshire              154
New Jersey                 152
New Mexico                 145
New York                   155
North Carolina             153
North Dakota               145
Northern Mariana Islands   128
Ohio                       147
Oklahoma                   150
Oregon                     157
Pennsylvania               150
Puerto Rico                143
Rhode Island               155
South Carolina             150
South Dakota               146
Tennessee                  151
Texas                      173
Utah                       160
Vermont                    149
Virgin Islands             142
Virginia                   149
Washington                 195
West Virginia              139
Wisconsin                  180
Wyoming                    145

This looks a little complex as we are *chaining* together a series of method
calls on the DataFrame. To understand this let's go from left to right and
unpack the statement in the order that the Python interpreter would.

- `df[['state', 'date']]` would return a data frame that contains only the
  variables `state` and `date`
- `df[['state', 'date']].groupby(by='state)` calls the `groupby` method on this small DataFrame. The `groupby` method takes an argument `by`, which we pass the value `state` into. This results in grouping records in the DataFrame that have the same value for `state` together.
- `df[['state', 'date']].groupby(by='state).count()` then says to aggregate
  the resulting DataFrame by counting the number of times each unique value of
  `state` appears.
  
We can contrast this single line of code with finding these values in your preferred spreadsheet software. 

### Ordinal

When a variable is recorded in an ordinal scale, the values imply an ordering
to the quantities. We do not have any ordinal scaled variables in the current
DataFrame, but if we use the `rank` method we can create one. Let's continue on
with our method chaining to rank the frequencies for the states:

In [23]:
df[['state', 'date']].groupby(by='state').count().rank()

date
state                         
Alabama                    6.5
Alaska                     9.5
Arizona                   52.0
Arkansas                  13.5
California                53.0
Colorado                  36.5
Connecticut               21.5
Delaware                  13.5
District of Columbia      25.0
Florida                   44.0
Georgia                   41.5
Guam                       3.0
Hawaii                    31.0
Idaho                      6.5
Illinois                  54.0
Indiana                   31.0
Iowa                      21.5
Kansas                    25.0
Kentucky                  31.0
Louisiana                 19.5
Maine                      9.5
Maryland                  36.5
Massachusetts             51.0
Michigan                  17.5
Minnesota                 31.0
Mississippi               13.5
Missouri                  25.0
Montana                    6.5
Nebraska                  48.0
Nevada                    36.5
New Hampshire             41.5
New Jersey                39.0
New Mexico                13.5
New York                  44.0
North Carolina            40.0
North Dakota              13.5
Northern Mariana Islands   1.0
Ohio                      19.5
Oklahoma                  31.0
Oregon                    46.0
Pennsylvania              31.0
Puerto Rico                6.5
Rhode Island              44.0
South Carolina            31.0
South Dakota              17.5
Tennessee                 36.5
Texas                     49.0
Utah                      47.0
Vermont                   25.0
Virgin Islands             4.0
Virginia                  25.0
Washington                55.0
West Virginia              2.0
Wisconsin                 50.0
Wyoming                   13.5

A couple of things should be noted about the `rank` method. First, the ranks
are ascending - that is the record with the lowest frequency will have the
rank 1. Second, in cases where two or more records have the same frequency, the
tie is reflected in a shared rank. We see this, for example, with Arkansas and
Delaware that are assigned the rank 13.5. Finally, the value of a rank denotes
ordering but not magnitude. This means we can say that California appeared more
frequently than Colorado, but we cannot  say that `53-36.5` is a meaningful difference.



### Interval and Ratio
The last levels of the measurement scale are interval and ratio. These are
closely related in that they share the characteristic that they both use
equal intervals that have meaning. The difference between interval and ratio
scales is that the ratio scale includes a zero point with the meaning: "does
not exist". For example, weight, height, and ruler measurements are on a ratio
scale, whereas time on a clock with hands is on an interval scale. In the
former cases, the 0 has a meaningful interpretation, but in the latter it does
not.

Because interval (and ratio) variables are measured on an interval scale, the
values can be subject to addition and subtraction. However, interval data
cannot be multiplied or divided (in a meaningful sense), while ratio data can
be. In other words, for a variable measured on the interval (or ratio) scale,
the difference between values is meaningful. So, while the difference in the
ranks of the frequencies for the states in the data is not meaningful (since
they are ordinal), the difference in the frequencies is. In other words, we can
say that California has appeared 40 more days in the data set than Colorado.
The frequencies are measured on the ratio scale, so we can meaningfully say
that California has appeared `1.27=191/151` as many times as Colorado in the data set 


## Data Set Types
As we have said, a **data set** is a collection of **observations** recorded
for individual **units** on a set of **variables**. The observations can be
recorded in a number of different ways, which gives rise to different **types**
of data sets. As we will see in later chapters of this book, different data set
types contain different levels of information which, in turn, allow for
different types of research questions to be formulated.

An important distinction draws from the definition of the unit in a data set.
In some data sets the observations could be considered *micro* in the sense
that they are recorded for individual people. In other cases, the unit
is an aggregation of individual values, such as when using
census data where information is reported for enumeration units such as tracts
or blocks. These are *macro* units.

In addition to the macro versus micro distinction, we can also classify a data
set as classified as *cross sectional*, *time series*, or *panel*.


### Cross Sectional
In cross sectional data sets, the observations are measured for a collection of units at one point in time for each variable.
For example, we can create a cross sectional DataFrame using pandas by
specifying a query to extract all records from the same date:

In [24]:
df_0802 = df[df.date == '2020-08-02']
df_0802.head()

date       state  fips   cases  deaths
8374  2020-08-02     Alabama     1   91444    1627
8375  2020-08-02      Alaska     2    3982      22
8376  2020-08-02     Arizona     4  178473    3769
8377  2020-08-02    Arkansas     5   43810     464
8378  2020-08-02  California     6  515937    9399

In [25]:
df_0802.tail()

date          state  fips  cases  deaths
8424  2020-08-02       Virginia    51  91782    2218
8425  2020-08-02     Washington    53  60161    1680
8426  2020-08-02  West Virginia    54   6854     117
8427  2020-08-02      Wisconsin    55  58990     956
8428  2020-08-02        Wyoming    56   2808      26

In [26]:
df_0802.shape

(55, 5)

So there were 55 cross sectional observations on this date.



### Time Series
In time series data sets, observations are measured for the same unit over
different periods of time for each variable. Using pandas, we can create a time
series DataFrame for California by changing the query:

In [27]:
df_cal = df[df.state == 'California']
df_cal.head()

date       state  fips  cases  deaths
5   2020-01-25  California     6      1       0
9   2020-01-26  California     6      2       0
13  2020-01-27  California     6      2       0
17  2020-01-28  California     6      2       0
21  2020-01-29  California     6      2       0

In [28]:
df_cal.shape

(191, 5)

So California is in the larger DF for 191 days. Since we have created this
smaller, state-specific, DF, let's do some exploration. First, what was the
maximum of the number of deaths in the state:

In [29]:
df_cal.deaths.max()

9399

When did the state reach that level of deaths?

In [30]:
df_cal[df.deaths==df_cal.deaths.max()]

/home/serge/anaconda3/envs/qm4pp/lib/python3.7/site-packages/ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


date       state  fips   cases  deaths
8378  2020-08-02  California     6  515937    9399

<!-- line 435 throws a warning in my notebook; it might be worth distinguishing between a traceback and a warning in an outside reference sheet> <-->
The cases and deaths are cumulative, so it makes sense that the maximum number
of cumulative deaths is reached on the last day of the sample. What about daily cases and deaths?

In [31]:
df_cal[['cases', 'deaths']].diff()

cases  deaths
5         NaN     NaN
9         1.0     0.0
13        0.0     0.0
17        0.0     0.0
21        0.0     0.0
...       ...     ...
8158  12904.0   192.0
8213   6414.0   101.0
8268   8004.0   213.0
8323   7234.0   143.0
8378   6430.0    34.0

[191 rows x 2 columns]

In [32]:
df_cal[['cases_d', 'deaths_d']] = df_cal[['cases', 'deaths']].diff()
df_cal.fillna(0, inplace=True)
df_cal.head()

/home/serge/anaconda3/envs/qm4pp/lib/python3.7/site-packages/pandas/core/frame.py:3062: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[k1] = value[k2]
/home/serge/anaconda3/envs/qm4pp/lib/python3.7/site-packages/pandas/core/frame.py:4317: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  downcast=downcast,


date       state  fips  cases  deaths  cases_d  deaths_d
5   2020-01-25  California     6      1       0      0.0       0.0
9   2020-01-26  California     6      2       0      1.0       0.0
13  2020-01-27  California     6      2       0      0.0       0.0
17  2020-01-28  California     6      2       0      0.0       0.0
21  2020-01-29  California     6      2       0      0.0       0.0

Here we have used the `diff` method to calculate the difference between the
cumulative cases
and deaths on consecutive dates. Then we assign these values into a new column in our dataframe
and replace the initial values with 0 (since we can't do a diff for the first observation).

So now we can find the date of the maximum daily reported deaths in the state:

In [33]:
df_cal[df_cal.deaths_d==df_cal.deaths_d.max()]

date       state  fips   cases  deaths  cases_d  deaths_d
8268  2020-07-31  California     6  502273    9222   8004.0     213.0

and the date of the maximum number of reported cases:

In [34]:
df_cal[df_cal.cases_d==df_cal.cases_d.max()]

date       state  fips   cases  deaths  cases_d  deaths_d
8158  2020-07-29  California     6  487855    8908  12904.0     192.0

### Cross Sectional - Time Series
The final structure for a data set combines cross sectional and time series
observations. This is sometimes referred to as *panel* data (and is the
motivation for the name of the *pandas* package). Two different cases are
typically encountered in practice. In the first, the panel is *balanced*,
meaning the same cross sectional observations appear in each of the time
periods. In an *unbalanced panel*, the number of cross sectional observations
is not constant across the time periods. 
<!-- Maybe a note here on which is more common and the implications of that? I expect unbalanced is more common, meaning that inferring data (as we do later in the chapter) could be further emphasized as a valuable skill here.<--> 

Our COVID data is an example of an unbalanced panel, since individual states enter
the data set at different points in time.


The distinctions between cross sectional, time series, and panel data sets are
important in terms of the statistical approaches that we will utilize. This is
because each type of data set conveys different types of variation that
requires specific types of analytical methods. For example, in the
cross-sectional DataFrame we created above, we would be able to examine the
variation in COVID cases and deaths *across states at the same date in time*.
In the time series example, our interest would be in the *variation in COVID
cases and deaths  within the state of California over time*. In the panel data
set, we have *variation over space  and time in COVID cases*.
<!-- this, to me, implies that panel data is prefered over the other types, but it might be worth stating explicitly that diff-in-diff analysis as enabled by panel data is a popular (standard?) policy evaluation method.><-->

## Data Manipulation


### Creating New Variables
Our DataFrame has two numerical variables that are of interest, the number of
cases and the number of deaths for each state and time period. While these are
useful, comparison of COVID trends between the different states is complicated
by the fact that the states vary substantially in their population sizes, as
well as demographic and ethnic composition, among other factors. In these cases, we would like to
develop a new variable that, in part, allows for a more standardized comparison
across states. 

One such variable is the death-to-case ratio which is the number of deaths
attributed to a particular disease during a time period divided by the number
of new cases identified during that same period. 

This is a ratio and not necessarily a proportion as some of the deaths may be
among individuals who developed the disease in an earlier period (and are thus
not in the denominator).


With pandas, we can create this new variable and add it to our DataFrame as
follows:

In [35]:
df['dtc100'] = df.deaths / df.cases * 100
df.head()

date       state  fips  cases  deaths  dtc100
0  2020-01-21  Washington    53      1       0     0.0
1  2020-01-22  Washington    53      1       0     0.0
2  2020-01-23  Washington    53      1       0     0.0
3  2020-01-24    Illinois    17      1       0     0.0
4  2020-01-24  Washington    53      1       0     0.0

Expressed per 100 cases, we see that early on there were no deaths, but at the
end of the sample the situation is much more dramatic:

In [36]:
df.tail()

date          state  fips  cases  deaths    dtc100
8424  2020-08-02       Virginia    51  91782    2218  2.416596
8425  2020-08-02     Washington    53  60161    1680  2.792507
8426  2020-08-02  West Virginia    54   6854     117  1.707032
8427  2020-08-02      Wisconsin    55  58990     956  1.620614
8428  2020-08-02        Wyoming    56   2808      26  0.925926

## Reading Remote Data

Thus far we have been working with a DataFrame that was generated from a local
file. That is, a file stored on our hard drive. Increasingly, data sets are
being made available on publicly accessible web sites, in which case being
able to read those sources into our workflow is an important skill. Here we
show how to use pandas together with a few other Python libraries to access the
original source directly:

In [37]:
import pandas as pd
import io
import requests
url = 'https://raw.githubusercontent.com/nytimes/covid-19-data/eb34a40dda05b211bdd24d1889ae1fcfe851984e/us-states.csv'
s=requests.get(url).content
remote_df=pd.read_csv(io.StringIO(s.decode('utf-8')))
remote_df.head()

date       state  fips  cases  deaths
0  2020-01-21  Washington    53      1       0
1  2020-01-22  Washington    53      1       0
2  2020-01-23  Washington    53      1       0
3  2020-01-24    Illinois    17      1       0
4  2020-01-24  Washington    53      1       0

We can use this remote reading capability to pull in additional variables to
further standardize the COVID activity by state. To control for state
population size, we can find an  on-line site such as:

In [38]:
from IPython.display import IFrame    
IFrame('https://simple.wikipedia.org/w/index.php?title=List_of_U.S._states_by_population&oldid=7015653', width=800, height=650)

We are using another Python module, `IPython` to embed an external live webpage
inside an
[Iframe](https://www.w3schools.com/html/html_iframe.asp). The Iframe allows us
to interact with the webpage as if we were visiting it, but we remain inside
our notebook environment. If we scroll around on the webpage we will see a data
table that lists, among other variables, `Population estiamtes, July 1, 2019`.
This is the variable we would like to integrate with our COVID data frame.

We could manually enter the values in that column into a python list and then
add that to the DataFrame. That would work, but would be error prone, and
tedious.

```{admonition} Habits
A good rule of thumb is that if we find ourselves in a situation that
requires us to repeat a task (such as entering a column of numbers into a
file), we should think about automating that task in a script. In doing so, we
should acknowledge that developing the script may actually require more time
than it would have taken if we typed the values in by hand. However, chances
are very good that we will have a reason to read data from websites again at
some point in our research careers, and thus the skills learned along the way
will save us time (and substantial time at that) over the course of our careers. In some cases, you'll be able to directly repurpose your old script to fit new purposes.
```

Ok, so let's think about how to automate the extraction of the population data
from the website. The challenge is that the data is buried inside quite a
bit of html code. Fortunately, pandas has a `read_html` method that will scrape
such a web page and extract any data tables.

In [39]:
url = "https://simple.wikipedia.org/w/index.php?title=List_of_U.S._states_by_population&oldid=7015653" # specific date
dfs = pandas.read_html(url)
dfs[0].head()

Rank in states & territories, 2019  Rank in states & territories, 2010  \
0                                 1.0                                 NaN   
1                                 NaN                                 NaN   
2                                 NaN                                 NaN   
3                                 NaN                                 NaN   
4                                 NaN                                 NaN   

          State  Population estimate, July 1, 2019[2]  \
0    California                              39512223   
1         Texas                              28995881   
2       Florida                              21477737   
3      New York                              19453561   
4  Pennsylvania                              12801989   

  Census population, April 1, 2010[3] Percent change, 2010–2019[note 1]  \
0                            37254523                              6.1%   
1                            25145561                             15.3%   
2                            18801310                             14.2%   
3                            19378102                              0.4%   
4                            12702379                              0.8%   

  Absolute change, 2010-2019  \
0                 +2,257,700   
1                 +3,850,320   
2                 +2,676,427   
3                    +75,459   
4                    +99,610   

  Total seats in the U.S. House of Representatives, 2013–2023  \
0                                                NaN            
1                                                NaN            
2                                                NaN            
3                                                NaN            
4                                                NaN            

  Estimated population per electoral vote, 2019[note 2]  \
0                                             718404      
1                                             763050      
2                                             740611      
3                                             670812      
4                                             640099      

  Estimated population per House seat, 2019  \
0                                    745514   
1                                    805441   
2                                    795472   
3                                    720502   
4                                    711222   

  Census population per House seat, 2010  \
0                                 702885   
1                                 698503   
2                                 696468   
3                                 717707   
4                                 705715   

  Percent of the total U.S. population, 2018[note 3]  
0                                             11.96%  
1                                              8.68%  
2                                              6.44%  
3                                              5.91%  
4                                              3.87%

<!-- It might be useful to introduce the notion of commenting in python early on to allow for easier annotation of each specific line. One of the major hangups I encountered when first learning python was trying to analyze what each line of the code was doing. Knowing that you were assigning the URL into a varible to pass into the read function wouldn't have made a ton of sense to me, but at least I would've understood that's what was happening.><-->

The `read_html` function returns a `list` of DataFrames - one DataFrame for
each table it encountered in the web page. In our case there is only one table
embedded in the page, which, because Python is a zero-offset language, we call with [0]. Using the `head` method on this DataFrame, we see that we
have extracted the general structure of the data table, and we can access our
variable of interest in this DataFrame:

In [40]:
pop_df = dfs[0]
pop_df['Population estimate, July 1, 2019[2]']

0      39512223
1      28995881
2      21477737
3      19453561
4      12801989
5      12671821
6      11689100
7      10617423
8      10488084
9       9986857
10      8882190
11      8535519
12      7614893
13      7278717
14      6949503
15      6833174
16      6732219
17      6137428
18      6045680
19      5822434
20      5758736
21      5639632
22      5148714
23      4903185
24      4648794
25      4467673
26      4217737
27      3956971
28      3565287
29      3205958
30      3155070
31      3193694
32      3080156
33      3017825
34      2976149
35      2913314
36      2096829
37      1934408
38      1787065
39      1792147
40      1415872
41      1359711
42      1344212
43      1068778
44      1059361
45       973764
46       884659
47       762062
48       731545
49       705749
50       623989
51       578759
52       165718
53       104914
54        55641
55        55194
56    325386357
57    327533795
58    328300544
59    331814684
Name: Population estimate, July 1, 2019[

What we want to do now is combine the population data from this DataFrame with our
covid data. To make our life easier (that is what scripting is all about),
let's change the name of this attribute:

In [41]:
pop_df.rename(columns={'Population estimate, July 1, 2019[2]': 'population'}, inplace=True)
pop_df.head()

Rank in states & territories, 2019  Rank in states & territories, 2010  \
0                                 1.0                                 NaN   
1                                 NaN                                 NaN   
2                                 NaN                                 NaN   
3                                 NaN                                 NaN   
4                                 NaN                                 NaN   

          State  population Census population, April 1, 2010[3]  \
0    California    39512223                            37254523   
1         Texas    28995881                            25145561   
2       Florida    21477737                            18801310   
3      New York    19453561                            19378102   
4  Pennsylvania    12801989                            12702379   

  Percent change, 2010–2019[note 1] Absolute change, 2010-2019  \
0                              6.1%                 +2,257,700   
1                             15.3%                 +3,850,320   
2                             14.2%                 +2,676,427   
3                              0.4%                    +75,459   
4                              0.8%                    +99,610   

  Total seats in the U.S. House of Representatives, 2013–2023  \
0                                                NaN            
1                                                NaN            
2                                                NaN            
3                                                NaN            
4                                                NaN            

  Estimated population per electoral vote, 2019[note 2]  \
0                                             718404      
1                                             763050      
2                                             740611      
3                                             670812      
4                                             640099      

  Estimated population per House seat, 2019  \
0                                    745514   
1                                    805441   
2                                    795472   
3                                    720502   
4                                    711222   

  Census population per House seat, 2010  \
0                                 702885   
1                                 698503   
2                                 696468   
3                                 717707   
4                                 705715   

  Percent of the total U.S. population, 2018[note 3]  
0                                             11.96%  
1                                              8.68%  
2                                              6.44%  
3                                              5.91%  
4                                              3.87%

Again, we see the power of the DataFrame, as its `rename` method takes a `columns`
keyword argument that accepts a `dict` specifying a pair of strings - the
original attribute name, and the replacement string. Here, we are only
renaming a single attribute, but if we wanted to we could add more entries to
the dictionary to rename additional attributes.

Let's further clean our DataFrame by shortening it to only include the state name and
our population variable:

In [42]:
pop_df = pop_df[['State', 'population']]
pop_df.head()

State  population
0    California    39512223
1         Texas    28995881
2       Florida    21477737
3      New York    19453561
4  Pennsylvania    12801989

Including the state column will make the next step easier. 


## Table Joins
To integrate our new `population` attribute from the `pop_df` DataFrame into our COVID
DataFrame, we are going to use a _table join_.  A join requires a _source_ and a
_target_ DataFrame. Here the `pop_df` is the source and our COVID DataFrame is the target. In
each DataFrame we need to have an attribute that we can use to match values of records from the
source DataFrame with corresponding records in the target DataFrame.

We will first try out the join which is accomplished using the `pandas`
function `merge`:

In [43]:
pandas.merge(df, pop_df, how='inner', left_on='state',
            right_on='State')

date                     state  fips  cases  deaths    dtc100  \
0     2020-01-21                Washington    53      1       0  0.000000   
1     2020-01-22                Washington    53      1       0  0.000000   
2     2020-01-23                Washington    53      1       0  0.000000   
3     2020-01-24                Washington    53      1       0  0.000000   
4     2020-01-25                Washington    53      1       0  0.000000   
...          ...                       ...   ...    ...     ...       ...   
8282  2020-07-29  Northern Mariana Islands    69     40       2  5.000000   
8283  2020-07-30  Northern Mariana Islands    69     42       2  4.761905   
8284  2020-07-31  Northern Mariana Islands    69     42       2  4.761905   
8285  2020-08-01  Northern Mariana Islands    69     44       2  4.545455   
8286  2020-08-02  Northern Mariana Islands    69     45       2  4.444444   

                         State  population  
0                   Washington     7614893  
1                   Washington     7614893  
2                   Washington     7614893  
3                   Washington     7614893  
4                   Washington     7614893  
...                        ...         ...  
8282  Northern Mariana Islands       55194  
8283  Northern Mariana Islands       55194  
8284  Northern Mariana Islands       55194  
8285  Northern Mariana Islands       55194  
8286  Northern Mariana Islands       55194  

[8287 rows x 8 columns]

which returns a new DataFrame resulting from the joined DataFrames. By scrolling to the right, we see
that the `State` and `population` values from the source DataFrame have been
added to the correct rows in our target DataFrame. To better understand what is going
on, let's unpack the `merge` function call which took five arguments: 

1. `df`: the _left_ or _target_ DataFrame
2. `pop_df`: the _right_ or _source_ DataFrame
3. `how`: a keyword argument specifying the type of join (here `inner`)
4. `left_on`: a keyword argument specifying the name of the attribute in the
   left DataFrame to join on
5. `right_on`: a keyword argument specifying the name of the attribute in the
   right DataFrame to join on


We have tried out the merge to see if it works. Now that it does, we can assign
it to reuse our original DataFrame name:

In [44]:
df = pandas.merge(df, pop_df, how='inner', left_on='state',
            right_on='State')
df.head()

date       state  fips  cases  deaths  dtc100       State  population
0  2020-01-21  Washington    53      1       0     0.0  Washington     7614893
1  2020-01-22  Washington    53      1       0     0.0  Washington     7614893
2  2020-01-23  Washington    53      1       0     0.0  Washington     7614893
3  2020-01-24  Washington    53      1       0     0.0  Washington     7614893
4  2020-01-25  Washington    53      1       0     0.0  Washington     7614893

This is very handy as it solves a
number of challenges for us. First, in the source DataFrame, the States do not appear
in the same order as in the target DataFrame. Second, in the target DataFrame, each state
appears multiple types, but only once in the source DataFrame. This latter is known as
a _one-to-many_ relationship as one record in the source DataFrame can match many
records in the target DataFrame. Third, the name of the attributes that we use in the
join are different in the two DataFrame (`State` vs. `state`).

````{margin}
```{note} 
For further details on table joins with pandas see the [official documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.merge.html?highlight=merge#pandas.merge).
```
````

Now that we have integrated the `population` data with our COVID data, we can
construct a new attribute: deaths per 100k population:

In [45]:
df['deaths100k'] = df.deaths / df.population * 10**5
df.head()


date       state  fips  cases  deaths  dtc100       State  \
0  2020-01-21  Washington    53      1       0     0.0  Washington   
1  2020-01-22  Washington    53      1       0     0.0  Washington   
2  2020-01-23  Washington    53      1       0     0.0  Washington   
3  2020-01-24  Washington    53      1       0     0.0  Washington   
4  2020-01-25  Washington    53      1       0     0.0  Washington   

   population  deaths100k  
0     7614893         0.0  
1     7614893         0.0  
2     7614893         0.0  
3     7614893         0.0  
4     7614893         0.0

We also can drop one of the redundant state variables that came with the merge:

In [46]:
df.drop(['State'], axis=1, inplace=True)
df.head()


date       state  fips  cases  deaths  dtc100  population  deaths100k
0  2020-01-21  Washington    53      1       0     0.0     7614893         0.0
1  2020-01-22  Washington    53      1       0     0.0     7614893         0.0
2  2020-01-23  Washington    53      1       0     0.0     7614893         0.0
3  2020-01-24  Washington    53      1       0     0.0     7614893         0.0
4  2020-01-25  Washington    53      1       0     0.0     7614893         0.0

## Saving DataFrames
We have come a long way in this chapter and the result is an integrated and
cleaned DF that is ready for further statistical analysis. We could of course,
rerun the code in this chapter to recreate this DF anytime that we need to.
This is one of the attractive features of computational scripting - being able
to reproduce our workflow. However, if we save our constructed data to an
external file, we can read that in during later analytical workflows and thus
avoid having to recreate the DF.

We will save our cleaned DF to a comma separated values (CSV) file:

In [47]:
df.to_csv('data/covid/covid_combined.csv', index=False)

This creates a new file that we could read in:

In [48]:
new_df = pandas.read_csv('data/covid/covid_combined.csv')
new_df.head()

date       state  fips  cases  deaths  dtc100  population  deaths100k
0  2020-01-21  Washington    53      1       0     0.0     7614893         0.0
1  2020-01-22  Washington    53      1       0     0.0     7614893         0.0
2  2020-01-23  Washington    53      1       0     0.0     7614893         0.0
3  2020-01-24  Washington    53      1       0     0.0     7614893         0.0
4  2020-01-25  Washington    53      1       0     0.0     7614893         0.0

And now we can move on to methods to visualize this data in the next chapter.



## Exercises

1. What was the date when California reported its 100th COVID death?

2. Generalize your code that you used for exercise 1 in the form of a function
that takes two arguments: `state` and `deaths` so that a user can find out the
date that any particular state reached a particular death total.  

3. Generate a DataFrame that reports the date and value that each state reached its
   maximum (peak) `dtc100` value.

4. In the `pop_df` DataFrame that was generated by scrapping an external web site, the
   column `Rank in states & territories, 2019` has a number of `NaN` values.
   What do these value represent and why are they occurring?